crates/iceberg/src/table.rs

```diff
@@ -16,28 +16,156 @@
 // under the License.
 
 //! Table API for Apache Iceberg
-use typed_builder::TypedBuilder;
+
+use std::sync::Arc;
 
 use crate::arrow::ArrowReaderBuilder;
+use crate::io::object_cache::ObjectCache;
 use crate::io::FileIO;
 use crate::scan::TableScanBuilder;
 use crate::spec::{TableMetadata, TableMetadataRef};
-use crate::{Result, TableIdent};
+use crate::{Error, ErrorKind, Result, TableIdent};
+
+/// Builder to create table scan.
+pub struct TableBuilder {
+    file_io: Option<FileIO>,
+    metadata_location: Option<String>,
+    metadata: Option<TableMetadataRef>,
+    identifier: Option<TableIdent>,
+    readonly: bool,
+    disable_cache: bool,
+    cache_size_bytes: Option<u64>,
+}
+
+impl TableBuilder {
+    pub(crate) fn new() -> Self {
+        Self {
+            file_io: None,
+            metadata_location: None,
+            metadata: None,
+            identifier: None,
+            readonly: false,
+            disable_cache: false,
+            cache_size_bytes: None,
+        }
+    }
+
+    /// required - sets the necessary FileIO to use for the table
+    pub fn file_io(mut self, file_io: FileIO) -> Self {
+        self.file_io = Some(file_io);
+        self
+    }
+
+    /// optional - sets the tables metadata location
+    pub fn metadata_location<T: Into<String>>(mut self, metadata_location: T) -> Self {
+        self.metadata_location = Some(metadata_location.into());
+        self
+    }
+
+    /// required - passes in the TableMetadata to use for the Table
+    pub fn metadata<T: Into<TableMetadataRef>>(mut self, metadata: T) -> Self {
+        self.metadata = Some(metadata.into());
+        self
+    }
+
+    /// required - passes in the TableIdent to use for the Table
+    pub fn identifier(mut self, identifier: TableIdent) -> Self {
+        self.identifier = Some(identifier);
+        self
+    }
+
+    /// specifies if the Table is readonly or not (default not)
+    pub fn readonly(mut self, readonly: bool) -> Self {
+        self.readonly = readonly;
+        self
+    }
+
+    /// specifies if the Table's metadata cache will be disabled,
+    /// so that reads of Manifests and ManifestLists will never
+    /// get cached.
+    pub fn disable_cache(mut self) -> Self {
+        self.disable_cache = true;
+        self
+    }
+
+    /// optionally set a non-default metadata cache size
+    pub fn cache_size_bytes(mut self, cache_size_bytes: u64) -> Self {
+        self.cache_size_bytes = Some(cache_size_bytes);
+        self
+    }
```

Comment:
This is the area I want to improve. If we can incorporate it into `FileIO`, we might avoid exposing the API wherever we need caching.

However, I think it's not a blocker of this PR. I'm willing to help improve this later.

Reply:
I like the sound of this! Look forwards to working with you on this afterwards :-)